laravel/horizon
# [4.x] Prevent horizon:purge from killing too many processes

**Summary**

This fixes the problem described in #769. To recap: if there's more than one Laravel application on one server, each with their own Horizon instance, running `php artisan horizon:purge` for one of the instances will kill valid Horizon processes of other instances. 

**Workaround**

The issue is supposed to be alleviated by using a supervisor daemon which will restart all processes after they get killed, but 

1. it seems like a workaround
2. when you read the code behind horizon:purge, its intent seems to be just killing invalid processes, e.g. the $description is "Terminate any rogue Horizon processes". That's confusing and makes it hard to determine what's actually going on
3. the [documentation](https://laravel.com/docs/7.x/horizon#deploying-horizon) doesn't say the daemon is required, so in my case I have a sysadmin who doesn't want to install it and insists on fixing the underlying issue and the client company agrees with him.

**Proposed fix**

MasterSupervisor: the hash

My company gave me some time to work on fixing the issue. The fix is based around adding an extra bit of information to the master supervisor's base name. It's a 4-character hash, similar to the random one already appended to base name, but this one instead of being random is based on the current directory, so will remain constant for the current Laravel application.

ProcessInspector: determining current processes

Instead of just doing `pgrep -f [h]orizon` which will also find processes of other Laravel apps, we only look for processes that include the current master supervisor base name. All supervisors and workers include it, but master supervisors themselves don't, so ProcessInspector@current will not return all the results we need. That's why we now also have ProcessInspector@mastersWithoutSupervisors - it returns ids of master supervisor processes that don't have supervisor processes underneath them. It can't determine whether the masters are from current Laravel app or not, but they don't have the expected child processes, so this time it's not a problem.

**Testing the PR**

In order to see the described behavior before and after the PR, you can use the following steps:

- set up two Laravel apps with Horizon on the same machine and run Horizon on both of them
- find out the master supervisor PID (e.g. using `ps -fx`) of one of them and kill it with with `kill -9`
- start a new master supervisor in place of the one you killed (needed for the monitoring information to update)
- run `php artisan horizon:purge` on the app where you killed the master. Before the PR, it will kill not just the rouge processes of this instance, but also the valid processes of the other one. After the PR, it will just kill the rouge processes of the current instance

That's it, let me know what you think!
